### ISSUE TITLE
[Performance] Boxing in OpenTelemetryLogger

### ISSUE THREAD

**Issue by CodeBlanch** [open-telemetry/opentelemetry-dotnet#1834]
We have `State` as an object on `LogRecord`:
```csharp
    public sealed class LogRecord
    {
        public object State { get; }
    }
```

The main `ILogger` extensions (eg LogInformation, LogWarning, etc.) all flow through [a single method](https://github.com/dotnet/runtime/blob/fa06656c41947e22fc6efd909cce0a6a180f1078/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerExtensions.cs#L393) which uses [struct FormattedLogValues](https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/FormattedLogValues.cs#L16) to store the state.

So that `object State` will cause a lot of boxing for users.

There are solutions to this, but it would be breaking. So @reyang @cijothomas @alanwest are we open to breaking the log API in order to fix this?

**Comment by reyang** [comment]
We cannot break the public API (in this case, it is the SDK package) within version 1.x.
What we can do is to introduce a `StateStruct { get; }` which does not require boxing, and update `State { get; }` to retrieve the struct.

**Comment by CodeBlanch** [comment]
> We cannot break the public API (in this case, it is the SDK package) within version 1.x.

I figured you would say that 😄 

> What we can do is to introduce a StateStruct { get; } which does not require boxing, and update State { get; } to retrieve the struct.

I don't think that will work, because the ILogger interface is generic:

```csharp
public void Log<TState> (Microsoft.Extensions.Logging.LogLevel logLevel, Microsoft.Extensions.Logging.EventId eventId, TState state, Exception exception, Func<TState,Exception,string> formatter);
```

I have a couple ideas let me mess with it and I'll reply back. Could we unseal `LogRecord` if needed? That wouldn't be breaking would it?

**Comment by reyang** [comment]
> I don't think that will work, because the ILogger interface is generic

Looks like you're right. I will take a look at the code and see if I have other ideas.
 
> I have a couple ideas let me mess with it and I'll reply back. Could we unseal `LogRecord` if needed? That wouldn't be breaking would it?

Yes we could! Unseal a class is not considered as a breaking change 😄

**Comment by CodeBlanch** [comment]
@reyang OK I pushed a potential solution to #1833 check it out and let me know what you think. It's tricky you really have to process the `TState` right in the `Log` call otherwise you will end up boxing it. I chatted with @cijothomas a bit on Slack he said (more or less) the plan was to always support scope on `LogRecord` so I decided to add scope + parsed state + message features on there and some options to control them. By default the options give you the behavior in 1.0. If you turn `ParseStateValues` ON, we will no longer box the state but instead convert it into an `IReadOnlyList<KeyValuePair<string, object>>` in an allocation-free way (for the common case).
